client/dashboard/data/index.ts

```diff
@@ -317,39 +317,67 @@ export const updateSiteSettings = async ( siteIdOrSlug: string, data: Partial< S
 };
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-function fromRawSiteSettings( settings: any ): SiteSettings {
-	const blog_public = Number( settings.blog_public );
-	const wpcom_coming_soon = Number( settings.wpcom_public_coming_soon );
-	const wpcom_public_coming_soon = Number( settings.wpcom_public_coming_soon );
+function fromRawSiteSettings( rawSettings: any ): SiteSettings {
+	// Pluck out raw settings which don't map directly to a field in SiteSettings.
+	const {
+		blog_public: blogPublicRaw,
+		wpcom_coming_soon: wpcomComingSoonRaw,
+		wpcom_public_coming_soon: wpcomPublicComingSoonRaw,
+		wpcom_data_sharing_opt_out: wpcomDataSharingOptOutRaw,
+		...settings
+	} = rawSettings;
+
+	const blog_public = Number( blogPublicRaw );
+	const wpcom_coming_soon = Number( wpcomComingSoonRaw );
+	const wpcom_public_coming_soon = Number( wpcomPublicComingSoonRaw );
+	const wpcom_data_sharing_opt_out = Boolean( wpcomDataSharingOptOutRaw );
 
 	if ( wpcom_coming_soon === 1 || wpcom_public_coming_soon === 1 ) {
 		settings.wpcom_site_visibility = 'coming-soon';
+		settings.wpcom_discourage_search_engines = false;
 	} else if ( blog_public === -1 ) {
 		settings.wpcom_site_visibility = 'private';
+		settings.wpcom_discourage_search_engines = false;
 	} else {
 		settings.wpcom_site_visibility = 'public';
+		settings.wpcom_discourage_search_engines = blog_public === 0;
 	}
+
+	settings.wpcom_prevent_third_party_sharing = wpcom_data_sharing_opt_out;
+
 	return settings;
 }
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 function toRawSiteSettings( settings: Partial< SiteSettings > ): any {
-	const rawSettings = settings as any; // eslint-disable-line @typescript-eslint/no-explicit-any
-
-	const { wpcom_site_visibility } = settings;
+	// Pluck out settings which don't map directly to a field in the raw settings.
+	const {
+		wpcom_site_visibility,
+		wpcom_discourage_search_engines,
+		wpcom_prevent_third_party_sharing,
+		...rest
+	} = settings;
+	const rawSettings = rest as any; // eslint-disable-line @typescript-eslint/no-explicit-any
 
 	if ( wpcom_site_visibility !== undefined ) {
 		if ( wpcom_site_visibility === 'coming-soon' ) {
 			rawSettings.blog_public = 0;
 			rawSettings.wpcom_public_coming_soon = 1;
+			rawSettings.wpcom_data_sharing_opt_out = false;
 		} else if ( wpcom_site_visibility === 'private' ) {
 			rawSettings.blog_public = -1;
 			rawSettings.wpcom_public_coming_soon = 0;
+			rawSettings.wpcom_data_sharing_opt_out = false;
 		} else {
-			rawSettings.blog_public = 1;
+			rawSettings.blog_public = wpcom_discourage_search_engines ? 0 : 1;
 			rawSettings.wpcom_public_coming_soon = 0;
+			rawSettings.wpcom_data_sharing_opt_out = wpcom_prevent_third_party_sharing;
 		}
+
+		// Take opportunity, while the user is switching visibility settings, to disable the legacy coming soon setting.
+		rawSettings.wpcom_coming_soon = 0;
```

Comment:
I wonder why we didn't have this in the tests but the tests still pass 🤔

Reply:
Perhaps test assertions are too permissive. This `toMatchObject` assertion isn't exhaustive.

https://github.com/Automattic/wp-calypso/pull/103661/files#diff-bcc30c86ea5afa51809a2b24b7a0e010ffdac0b354f0ada145f7d8068f5445ffR61

I figured this component only cares about whether its settings get changed, it shouldn't care whether there's some other part of the code that does something else with the settings. But do you think this is bad form?

Reply:
> it shouldn't care whether there's some other part of the code that does something else with the settings.

But `wpcom_coming_soon`, `wpcom_discourage_search_engines`, `wpcom_data_sharing_opt_out` getting all sent from the form, is part of this form's logic, no? It's literally the form's responsibility (and expectation).

Reply:
Yes, now that `wpcom_coming_soon` has been added, it's clarified that it is this form's responsibility to ensure all these settings are set. When I do the mapping refactor I would want these tests to ensure none of this behaviour has changed :) So that's a good justification for why I should do a stricter assert 😃 

Fixed in 66b11634dba239fddefca705cb2c29f93f8e9e93